File: 1482-how-many-numbers-are-smaller-than-the-current-number/how-many-numbers-are-smaller-than-the-current-number.go
import (
	"slices"
)

func smallerNumbersThanCurrent(nums []int) []int {
	// create a temp copy of original array
	sortedNums := append([]int(nil), nums...)

	// sort temp
	slices.Sort(sortedNums)

	// create result array
	result := []int{}

	// create value index hash map
	valueIndexMap := make(map[int]int)

	for index, value := range sortedNums {
		if _, ok := valueIndexMap[value]; !ok {
			valueIndexMap[value] = index
		}
	}

	for _, value := range nums {
		result = append(result, valueIndexMap[value])
	}

	return result
}

File: 1482-how-many-numbers-are-smaller-than-the-current-number/how-many-numbers-are-smaller-than-the-current-number.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var smallerNumbersThanCurrent = function (nums) {
    const temp = [...nums];
    nums.sort((a, b) => a-b);
    const obj = {};
    const res = [];

    for (let i = 0; i < nums.length; i++) {
        if (!obj.hasOwnProperty(nums[i])) obj[nums[i]] = i;
    }

    for (const num of temp) {
        res.push(obj[num]);
    }

    return res;
};
